### PR TITLE
Generate manifest with static calls

### PR DESCRIPTION
**Pull request by shargon**
Close https://github.com/neo-project/neo-devpack-dotnet/issues/475

### CODE REVIEW

**Comment by superboyiii** [comment]
@shargon I'll test this later but I think it's better not be merged to RC1 since not much time left.

**Comment by erikzhang** [comment]
@superboyiii It should be added to RC1.

**Comment by superboyiii** [comment]
I've tested this. It works but I'm confused if it makes sense in this mechanism.
1. In previous, unnative contracts are allowed to be called by all contracts as default. And if add 1 permission, it becomes only allowed this permission method to call.
2. In this PR, unnative contracts are not allowed to be called by all contracts as default. That means if we'd like to call any method of other SC, we have to upgrade our SC again and again. In most DAPP, modules are added step by step not at the same time. In this way, I think most developers will add [ContractPermission("*", "*")] at the beginning and upgrade ContractPermission later. So is it useful we set forbidden all as the default?
@shargon @erikzhang

**Comment by shargon** [comment]
We can add a compiler flag `--allowAllCalls`?

**Comment by erikzhang** [comment]
```c#
[ContractPermission("*")]
```

You need to manually add this line to declare the calling permissions of all contracts.

**Comment by erikzhang** [comment]
Tested?

**Comment by superboyiii** [comment]
> Tested?

I will test it soon.

**Comment by superboyiii** [comment]
@shargon I set `[ContractPermission("*", "*")]` but `ToDelegate()` still return fault because of authorization. Does this mean if natvie management contract is allowed in permission than all permission must add manually? "*" seems doesn't work.

```
{
    "jsonrpc": "2.0",
    "id": 3,
    "result": {
        "script": "ABQMFI5SMl22gwrQirvz2CzRnQyvhFHfDBSabQINJN1EW/uACj2hZ5UvHKCtDwwUj0sLioGSzZgeFe060WuXVzteXjgUwB8MBXRlc3QyDBTX2B1RYQz8amWTYixFhXS9jndqD0FifVtS",
        "state": "FAULT",
        "gasconsumed": "3471120",
        "exception": "An unhandled exception was thrown. No authorization.",
        "stack": []
    }
}
```

**Comment by erikzhang** [comment]
> [ContractPermission("*", "*")]

I think it not `[ContractPermission("*", "*")]` but `[ContractPermission("*")]`.

**Comment by superboyiii** [comment]
> > [ContractPermission("*", "*")]
> 
> I think it's not `[ContractPermission("*", "*")]` but `[ContractPermission("*")]`.
I've tried both, `[ContractPermission("*", "*")] ` and `[ContractPermission("*")]` does no difference in either nef hash or manifest json result. And I saw my permission and native permission exist both, which should be the final result?

image

**Comment by erikzhang** [comment]
> which should be the final result?

The first one.

But I think we should optimize for it. @shargon

**Comment by superboyiii** [comment]
Sorry, I turned witnessScopes from `CalledByEntry` to `Global`, everything works well.
So in current mechanism, default still is `allowAll` and whatever your permission is, `contrtactmanagement` can always be invoked in `getContract`, `destroy`, `update`.

**Comment by erikzhang** [comment]
You still need the permissions. Otherwise it should fail.

**Comment by superboyiii** [comment]
> You still need the permissions. Otherwise it should fail.

If one permission is manually added, then will be `denyAll` except `contractmanagerment`. But if you don't set any permission, it's the same effect as `[ContractPermission("*", "*")]`.

**Comment by superboyiii** [comment]
Test **PASS**
It works well on mine.

**Review by superboyiii — DISMISSED** [review]

**Comment by erikzhang** [comment]
@shargon Can you optimize the permission list?

**Comment by shargon** [comment]
> Can you optimize the permission list?

Yes, I will

**Comment by superboyiii** [comment]
> > Can you optimize the permission list?
> 
> Yes, I will

If optimize it, we need do it quickly since I'll release CLI today and set up Testnet on 18th. It should be merged before Testnet.

**Comment by shargon** [comment]
@superboyiii please test it

**Comment by superboyiii** [comment]
> @superboyiii please test it

It works on mine.

**Review by superboyiii — DISMISSED** [review]

**Comment by superboyiii** [comment]
@erikzhang Need your review.

**Comment by erikzhang** [comment]
@shargon Please use the `PermissionBuilder` added in da3b922.

**Comment by shargon** [comment]
@superboyiii please test it

File: src/Neo.Compiler.MSIL/PermissionBuilder.cs [hunk 1]
```diff
@@ -44,7 +49,7 @@ public void Add(string hash, string method)
         public JArray ToJson()
         {
             JArray permissions = new();
-            if (isWildcard)
+            if (isWildcard || normalItems.Count == 0)
```
Comment (erikzhang) [inline src/Neo.Compiler.MSIL/PermissionBuilder.cs:52]: Why? I think `*, *` should be explicitly declared.

Comment (shargon) [reply src/Neo.Compiler.MSIL/PermissionBuilder.cs:52]: Because otherwise the manifest will be invalid, isn't it?

Comment (erikzhang) [reply src/Neo.Compiler.MSIL/PermissionBuilder.cs:52]:
```
"permissions": []
```

It's valid.

**Comment by superboyiii** [comment]
@shargon Double `[]`?
**
image
**

**Review by erikzhang — APPROVED** [review]

**Comment by superboyiii** [comment]
Test **PASS**
Works well on mine.

**Comment by erikzhang** [comment]
Please fix the UT.

**Comment by superboyiii** [comment]
@shargon I just push a PR on your repo to fix this, could you review it? https://github.com/neo-project/neo-modules/pull/564

**Comment by shargon** [comment]
@superboyiii merged!

**Comment by superboyiii** [comment]
@erikzhang Merge?

**Review by erikzhang — APPROVED** [review]

**Comment by erikzhang** [comment]
It's better to test it again before merging.

**Comment by superboyiii** [comment]
> It's better to test it again before merging.

Give me half an hour.

**Comment by superboyiii** [comment]
> It's better to test it again before merging.

Test **PASS**
Works well on mine.

**Review by superboyiii — APPROVED** [review]

**Review by erikzhang — APPROVED** [review]

**Pull request merged**